tamagotchi/src/contract.rs

```diff
@@ -0,0 +1,182 @@
+            tamagotchi.date_of_birth = tamagotchiAge;
+            msg::reply(TmgEvent::Age(tamagotchiAge), 0).expect("Errorin sending tamagotchi age");
+        },
+        TmgAction::Feed => {
+            tamagotchi.fed = tamagotchi_fed_updated(blocksHeight);
+            tamagotchi.fed_block = blocksHeight;
+            msg::reply(TmgEvent::Fed, 0).expect("Error sending tamagotchi variant 'Fed'");
+        },
+        TmgAction::Play => {
+            tamagotchi.entertained = tamagotchi_entertained_updated(blocksHeight);
+            tamagotchi.entertained_block = blocksHeight;
+            msg::reply(TmgEvent::Entertained, 0).expect("Error sending tamagotchi variant 'Entertained'");  
+        },
+        TmgAction::Sleep => {
+            tamagotchi.rested = tamagotchi_rested_updated(blocksHeight);
+            tamagotchi.rested_block = blocksHeight;
+            msg::reply(TmgEvent::Slept, 0).expect("Error sending tamagotchi variant 'Slept'");  
+        }
+    }
+}
+
+#[no_mangle]
+extern "C" fn state() {
+    let blocksHeight = blocks_height();
+    let tamagotchi = state_ref();
+    let state = Tamagotchi {
+        name: String::from(&tamagotchi.name),
+        date_of_birth: tamagotchi.date_of_birth,
+        owner: tamagotchi.owner.clone(),
+        fed: tamagotchi_total_fed(blocksHeight),
+        fed_block: tamagotchi.fed_block,
+        entertained: tamagotchi_total_entertained(blocksHeight),
+        entertained_block: tamagotchi.entertained_block,
+        rested: tamagotchi_total_rested(blocksHeight),
+        rested_block: tamagotchi.rested_block
+    };
+    msg::reply(tamagotchi, 0).expect("Failed to share state");
+}
+
+fn blocks_height() -> u64 {
+    u64::from(exec::block_height())
+}
+
+fn tamagotchi_total_fed(blocksHeight: u64) -> u64 {
+    let tamagotchi = state_ref();
+    let tamagotchi_hungry = (blocksHeight - tamagotchi.fed_block) * HUNGER_PER_BLOCK;
+    
+
+    if tamagotchi.fed > tamagotchi_hungry {
+        tamagotchi.fed - tamagotchi_hungry
+    } else {
+        0
+    }
+}
+
+fn tamagotchi_total_entertained(blocksHeight: u64) -> u64 {
+    let tamagotchi = state_ref();
+    let tamagotchi_bored = (blocksHeight - tamagotchi.entertained_block) * BOREDOM_PER_BLOCK;
+    
+    if tamagotchi.entertained > tamagotchi_bored {
+        tamagotchi.entertained - tamagotchi_bored
+    } else {
+        0
+    }
+}
+
+fn tamagotchi_total_rested(blocksHeight: u64) -> u64 {
+    let tamagotchi = state_ref();
+    let tamagotchi_tired = (blocksHeight - tamagotchi.rested_block) * ENERGY_PER_BLOCK;
+
+    if tamagotchi.rested > tamagotchi_tired {
+        tamagotchi.rested - tamagotchi_tired
+    } else {
+        0
+    }
+}
+
+fn tamagotchi_fed_updated(blocksHeight: u64) -> u64 {
+    let tamagotchi = state_ref();
+    let actual_fed = tamagotchi_total_fed(blocksHeight);
+    let total_fed = actual_fed + FILL_PER_FEED;
+
+    if total_fed <= 10_000 {
+        total_fed
+    } else {
+        10_000
+    }
```

Comment:
It might be shorter:
```suggestion
    total_fed.max(10_000)
```